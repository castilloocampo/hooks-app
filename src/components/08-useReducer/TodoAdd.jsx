import React from 'react';
import PropTypes from 'prop-types';
import useForm from '../../hooks/useForm';

const initialFormState = {
  description: ""
};

const TodoAdd = ({ onHandleAddTodo }) => {
  const [{ description }, handleInputChange, resetForm] = useForm(initialFormState);

  const handleSubmit = (e) => {
    e.preventDefault();
    if (description.trim() !== "") {
      const newTodo = {
        id: new Date().getTime(),
        description,
        done: false
      }
      onHandleAddTodo(newTodo);
      resetForm();
    }
  }


  return (
    <>
      <h4>Agregar TODO</h4>
      <hr />
      <form
        onSubmit={handleSubmit}
        data-testid="form-todoadd"
      >
        <input
          type="text"
          name="description"
          placeholder="Aprender..."
          autoComplete="off"
          value={description}
          onChange={handleInputChange}
          data-testid="input-todoadd"
        />

        <button
          type='submit'
          className='btn btn-outline-primary w-full'
          data-testid="button-todoadd"
        >
          Agregar
        </button>
      </form>
    </>
  );
};


TodoAdd.propTypes = {
  onHandleAddTodo: PropTypes.func.isRequired,
};


export default TodoAdd;
