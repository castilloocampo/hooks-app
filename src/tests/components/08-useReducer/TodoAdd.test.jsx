
import { fireEvent, render } from '@testing-library/react';
import TodoAdd from '../../../components/08-useReducer/TodoAdd';
import { screen } from '@testing-library/react';

describe('TodoAdd component', () => {
  const handleAddTodo = jest.fn();

  test('should show correctly', () => {

    const { container } = render(<TodoAdd
      onHandleAddTodo={handleAddTodo}
    />);

    expect(container).toMatchSnapshot();

  });

  test("shouldn't call onHandleAddTodo ", () => {
    render(<TodoAdd
      onHandleAddTodo={handleAddTodo}
    />);
    const btnSubmit = screen.getByTestId('button-todoadd');
    fireEvent.click(btnSubmit);

    expect(handleAddTodo).not.toBeCalled();
  });

  test("shouldn't call onHandleAddTodo ", () => {
    render(<TodoAdd
      onHandleAddTodo={handleAddTodo}
    />);
    const btnSubmit = screen.getByTestId('button-todoadd');
    fireEvent.click(btnSubmit);

    expect(handleAddTodo).not.toBeCalled();
  });

  test('should call onHandleAddTodo', () => {
    render(<TodoAdd
      onHandleAddTodo={handleAddTodo}
    />);
    const formSubmit = screen.queryByTestId('form-todoadd');
    const input = screen.getByTestId('input-todoadd');
    const value = "Learn SCRUM";
    fireEvent.change(input, { target: { value } });
    fireEvent.submit(formSubmit);
    expect(handleAddTodo).toHaveBeenCalled();
    expect(handleAddTodo).toHaveBeenCalledWith({
      id: expect.any(Number),
      description: value,
      done: false,
    });
    expect(input.getAttribute("value")).toBe("");
  });

});