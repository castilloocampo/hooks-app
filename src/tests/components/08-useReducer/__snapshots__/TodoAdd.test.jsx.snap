// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoAdd component should show correctly 1`] = `
<div>
  <h4>
    Agregar TODO
  </h4>
  <hr />
  <form
    data-testid="form-todoadd"
  >
    <input
      autocomplete="off"
      data-testid="input-todoadd"
      name="description"
      placeholder="Aprender..."
      type="text"
      value=""
    />
    <button
      class="btn btn-outline-primary w-full"
      data-testid="button-todoadd"
      type="submit"
    >
      Agregar
    </button>
  </form>
</div>
`;
